### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by weblaetitia**

### CODE REVIEW

File: src/CoverCrypt.tsx [hunk 1]
```diff
@@ -189,48 +275,141 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
+        localEncryptedData.map(async (row) => {
+          try {
+            const hr = await decryptDataLocally(row.hr, retrievedDecryptionKey.bytes());
+            const decryptedHR = JSON.parse(hr);
+            return decryptedHR;
+          } catch {
+            //
+          }
+        })
+      );
```
Comment (ccorsin) [inline src/CoverCrypt.tsx:344]: maybe refacto and create only one function to decrypt and get both marketing and HR objects, passing the right decryption function as an argument ?

File: src/CoverCrypt.tsx [hunk 2]
```diff
@@ -189,48 +275,141 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
+          salary: clearHR[key]?.salary != null ? clearHR[key].salary : "–",
+        };
+      });
+
+      setLocalClearData(clearEmployee as Employee[]);
+      setLocalEncryptedData(undefined);
+    }
+  };
+  const handleDecryptInKms = async (): Promise<void> => {
+    // BROWSER
```
Comment (ccorsin) [inline src/CoverCrypt.tsx:320]: Why is there a browser section in handleDecryptInKms ? :)

File: src/CoverCrypt.tsx [hunk 3]
```diff
@@ -189,48 +275,141 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
+          first: row?.first != null ? row.first : "–",
+          last: row?.last != null ? row.last : "–",
+          country: row?.country != null ? row.country : "–",
+          email: clearHR[key]?.email != null ? clearHR[key].email : "–",
+          salary: clearHR[key]?.salary != null ? clearHR[key].salary : "–",
+        };
+      });
+
+      setKmsClearData(clearEmployee as Employee[]);
+      setKmsEncryptedData(undefined);
```
Comment (ccorsin) [inline src/CoverCrypt.tsx:398]: keep only this + refacto ?

File: src/CoverCrypt.tsx [hunk 4]
```diff
@@ -277,75 +457,145 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
-              )}
+              {localClearData && <EmployeeTable caption={"Decrypted in browser"} data={localClearData} />}
 
               {/* ENCRYPT/DECRYPT IN KMS */}
-              <ButtonGroup isAttached variant="outline" isDisabled={keysId == null}>
+              <HeadingWithCode heading="Encrypt & Decrypt data in KMS" />
+              <Code>/src/actions/encryptDataInKms.ts</Code>
+              <CodeHigligter codeInput={code?.encryptDataInKms} />
+              <Code>/src/actions/decryptDataLocally.ts</Code>
+              <CodeHigligter codeInput={code?.decryptDataLocally} />
```
Comment (ccorsin) [inline src/CoverCrypt.tsx:587]: wrong elements - locally decryption, not in KMS

File: src/CoverCrypt.tsx [hunk 5]
```diff
@@ -277,75 +457,145 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
+              <ButtonGroup isAttached variant="outline" isDisabled={keyPair == null}>
                 <Button onClick={() => handleEncrypt({ browser: false })} width="50%">
                   Encrypt data in KMS
                 </Button>
-                <Button onClick={() => handleDecrypt({ browser: false })} width="50%" isDisabled={kmsEncryptedData == null}>
```
Comment (ccorsin) [inline src/CoverCrypt.tsx:334]: I can see you removed this refacto - why ?

File: src/Layout.tsx [hunk 6]
```diff
@@ -11,3 +33,119 @@ const Layout = ({ children }: { children: JSX.Element }): JSX.Element => {
...
+  );
+};
+
+export const Jsoncode = ({ code }: { code: unknown }): JSX.Element => {
+  return (
+    <Box bg="orange.50" border="1px" borderColor="orange.100" color="gray.700" fontSize="sm" p="5">
+      <pre style={{}}>{JSON.stringify(code, undefined, 2)}</pre>
+    </Box>
+  );
+};
```
Comment (ccorsin) [inline src/Layout.tsx:78]: I don't think this is used anymore - remove this ?

File: public/actions/decryptDataLocally.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { CoverCrypt, PrivateKey } from "cloudproof_js";
+
+//
+// Decrypt data locally
+//
+export const decryptDataLocally = async (encrypted: Uint8Array, decryptionKey: Uint8Array | PrivateKey): Promise<string> => {
```
Comment (ccorsin) [inline public/actions/decryptDataLocally.ts:6]: a detail - but to stay coherent with above : encrypted -> encryptedData ?

File: public/actions/decryptDataLocally.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  const { CoverCryptHybridDecryption } = await CoverCrypt();
+  const decrypter = new CoverCryptHybridDecryption(decryptionKey);
+  const decrypted = decrypter.decrypt(encrypted);
```
Comment (ccorsin) [inline public/actions/decryptDataLocally.ts:9]: decryptedData ?

File: public/actions/decryptDataInKms.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { KmsClient } from "cloudproof_js";
+import { BACKEND_URL } from "./backendConfig";
+
+//
+// Decrypt data in KMS
+//
+export const decryptDataInKms = async (encryptText: Uint8Array, kmsToken: string, decryptionKey: string): Promise<string> => {
+  const client = new KmsClient(BACKEND_URL, kmsToken);
+  const decrypted = await client.coverCryptDecrypt(decryptionKey, encryptText);
```
Comment (ccorsin) [inline public/actions/decryptDataInKms.ts:9]: decryptedData ?

File: public/actions/encryptDataLocally.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { CoverCrypt, Policy } from "cloudproof_js";
+
+//
+// Encrypt data locally
+//
+export const encryptDataLocally = async (
+  publicKey: Uint8Array,
+  policy: Policy,
+  accessPolicy: string,
+  dataToEncrypt: Uint8Array
```
Comment (ccorsin) [inline public/actions/encryptDataLocally.ts:10]: same - clearData ?

File: public/actions/retrieveDecryptionKey.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { KmsClient, PrivateKey } from "cloudproof_js";
+import { BACKEND_URL } from "./backendConfig";
+
+//
+// Retrieve Decryption Key
+//
+export const retrieveDecryptionKey = async (kmsToken: string, udkID: string): Promise<PrivateKey> => {
```
Comment (ccorsin) [inline public/actions/retrieveDecryptionKey.ts:7]: udkID not so explicite for sample code ?

File: public/actions/retriveDecryptionKey.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { KmsClient, PrivateKey } from "cloudproof_js";
+import { BACKEND_URL } from "./backendConfig";
+
+//
+// Retrieve Decryption Key
+//
+export const retrieveDecryptionKey = async (kmsToken: string, udkID: string): Promise<PrivateKey> => {
```
Comment (ccorsin) [inline public/actions/retriveDecryptionKey.ts:7]: same question about udkID

File: public/actions/retriveKeyPair.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import { KmsClient } from "cloudproof_js";
+import { BACKEND_URL } from "./backendConfig";
+import { KeyPair, KeysBytes } from "./types";
+
+//
+// Retrieve Key Pair
+//
+export const retrieveKeyPair = async (kmsToken: string, keysId: KeyPair): Promise<KeysBytes> => {
```
Comment (ccorsin) [inline public/actions/retriveKeyPair.ts:8]: Names are a bit confusing for me - why keysId has type KeyPair ? Are those keys or simple ids ?

File: src/Layout.tsx [hunk 14]
```diff
@@ -11,3 +33,119 @@ const Layout = ({ children }: { children: JSX.Element }): JSX.Element => {
...
+      <Heading as="h2" size="lg">
+        {heading}
+      </Heading>
+      {code && typeof code === "object" && code.map((item, i) => <Code key={i}>{item}</Code>)}
+      {code && typeof code === "string" && <Code>{code}</Code>}
+    </>
+  );
+};
+
+export const EmployeeTable: React.FC<{ data: Employee[]; caption?: string }> = ({ data, caption }) => {
```
Comment (ccorsin) [inline src/Layout.tsx:93]: why is this in Layout component ?

File: src/Layout.tsx [hunk 15]
```diff
@@ -11,3 +33,119 @@ const Layout = ({ children }: { children: JSX.Element }): JSX.Element => {
...
+            );
+          })}
+        </Tbody>
+        {caption && <TableCaption>{caption}</TableCaption>}
+      </Table>
+    </TableContainer>
+  );
+};
+
+export const EncryptedTable: React.FC<{ data: EncryptedResult[]; caption?: string }> = ({ data, caption }) => {
```
Comment (ccorsin) [inline src/Layout.tsx:122]: why is this in Layout component ?

File: src/CoverCrypt.tsx [hunk 16]
```diff
@@ -189,48 +275,141 @@ const CoverCrypt: React.FC<{ kmsToken: string }> = ({ kmsToken }) => {
...
-          );
-          setKmsClearData(clear);
-          setKmsEncryptedData(undefined);
-        }
-      }
-    } catch (error) {
-      toastError(error);
+  const handleDecryptLocally = async (): Promise<void> => {
+    if (localEncryptedData && decryptionKey) {
+      const retrievedDecryptionKey = await retrieveDecryptionKey(kmsToken, decryptionKey as string);
```
[the remark this answers is not in the document]
Comment (weblaetitia) [reply src/CoverCrypt.tsx:280]: Ok renamed, it retrieve a key from the key ID

File: src/CoverCrypt.tsx (hunk 5 above)
Comment (weblaetitia) [reply src/CoverCrypt.tsx:334]: because it was messy, 2 differents actions in same function

**Pull request merged**